rfcs/002-merge-types-terms-syntax.md

```diff
@@ -0,0 +1,452 @@
+design guideline, so we also restrain from giving a formal definition for `~`).
+
+1. For all `t: Term`, `term(!t) ~ t` and for all `T: Type`, `type(!T) = T` .
+   That is, a term or a type in the current syntax with `#` removed will be
+   interpreted the same in the new syntax.
+2. For all `T: Type`, `term(!T) ~ §T`. That is, the interpretation of something
+   that is a type in the current syntax as a term in the new syntax must behave
+   the same as the corresponding contract.
+
+### `term`
+
+The `term` translation mainly inserts implicit `§` automatically.
+
+- For most term-only constructs (that is everything excepted
+  records and constructors coming from types), `term` is an homomorphism (e.g.
+  for lists `term [t1, ..., tn] = [ term(t1), ..., term(tn)]`).
+- For type-only constructs (ground types, `->`, and `{_: T}`), it is
+  `term(T) = §(type(T))`, that is the contract associated to the type.
+
+#### Records
+
+It remains to treat the case of records, that can mix previously type and term
+constructs.
+
+- a row declaration `foo : Num` is translated as it is. This is actually already
+  supported since the piecewise signature feature. `{foo : Num}` is just
+  equivalent to `{foo | Num}` operationally when `foo` doesn't have a
+  definition.
+- For a tail, because of the predicate enforced in the grammar, we know that a
+  record `{ body; tail}` in `UniTerm` must satisfy `is_record_type({body})`. We
+  just convert it to a type, and derive the corresponding contract: `term({body;
+  tail}) = §type({body; tail})`.
+
+### `type`
+
+Dually to terms:
+
+- For type-only constructs, `type` is an homomorphism. For example,
+  `type(t -> t') = type(t) -> type(t')`.
+- For term-only constructs, we take the opaque type associated to a contract.
+  That is, `type(t) = #term(t)`. For example,
+  `type(fun x => body) = #term(fun x => body)`.
+
+#### Records
+
+As for terms, the case of records remains. For records that are actually the
+embedding of a `Type`, such as `{foo : Num, bar: {baz: Num}}`, the translation
+is obvious. For records that are the embedding of a `Term` without annotations,
+we can simply proceed as above: `type({foo = 1, bar = "baz"}) = #{foo = 1, bar =
+"baz"}`.
+
+The case for records that mixes the both `Term` and `Type` is less clear. Take
+for example: `x : {foo : Num = 1}`. Should we consider this as a totally opaque
+record, that is `type({foo : Num = 1}) = #{foo : Num = 1}` or take advantage
+from the information that `foo` is necessarily a number and set `type({foo : Num
+= 1}) = {foo: Num}`? The problem with the latter approach is that we loose the
+property of *blame safety*, which is that "well-typed program can't be blamed".
+Indeed, the extracted type is less precise that the actual contract. For
+example, `{foo = 2} : {foo : Num, foo = 1}` would pass typechecking but fails at
+runtime on the failure of merging `1` and `2`. There are also multiple
```

Comment:
It probably wouldn't, actually. With the proposed semantics, the runtime contract would presumably be `§{foo: Num}`.

It doesn't make it a good idea, though.

Reply:
Your remark makes me think about something. The `type()` translation erases metadata, so it may have some unwanted consequences, such as `{foo = 1} | {foo : Num | doc "Foo field"}` loosing its documentation at runtime (`type({foo : Num | doc "Foo field"}) = {foo : Num}`). Here maybe the contract feels slightly off, mixing type annotation and doc, but this would be even more natural if we set `type({foo | Num}) = {foo : Num}` at some point, as we would certainly write `{foo = 1} | #{foo | Num | doc "..."}`.

There's also some edgy details about `§{foo : Num}` and `#{foo | Num}` not behaving exactly the same right now. I see `type()` as extracting static type information from an annotation, but being quite irrelevant to runtime (as, at most, it will drop metadata, but never add anything helpful to evaluation). Thus, it may make sense to use the translation uniquely at typechecking time, to extract static type information, but keep the original input for runtime. I may had this in mind implicitly when I wrote this paragraph.

Reply:
It does sound tricky. At any rate, I agree with being as conservative as we can get away with. We can discover interesting usage patterns with experience.